### PR TITLE
Documentation about how to run Kubernetes in an hybrid env

### PR DESCRIPTION
**Pull request by jcorioland**
**What this PR does / why we need it**: adds documentation that explains considerations to have before using acs-engine to run a Kubernetes cluster in an hybrid cloud/on-premise environment

**If applicable**:
- [x] documentation

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -20,3 +20,4 @@ This cluster definition examples demonstrate how to create a customized Docker E
 * [Attached Disks](../examples/disks-storageaccount) - shows how to attach up to 4 disks per node
 * [Managed Disks](../examples/disks-managed) (under private preview) - shows how to use managed disks
 * [Large Clusters](../examples/largeclusters) - shows how to create cluster sizes of up to 1200 nodes
+* [Running Kubernetes in an hybrid environment](kubernetes/hybrid-environment.md) - considerations around how you will configure your cluster, and plan your deployment when running Kubernetes in a Cloud/On-Premise environment
```
Comment (CecileRobertMichon) [inline docs/README.md]: Should be `in a hybrid environment`

File: docs/kubernetes/hybrid-environment.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
+# Running Kubernetes in a hybrid environment
+
+## Introduction
+
+Setting up Kubernetes in a hybrid environment imposes some requirement on the setup, and implies some considerations around how you will configure your cluster, and plan your deployments. In this guide, we will look at a typical hybrid cloud setup to deploy Kubernetes.
+
+Note : We do not cover a Kubernetes cluster _spanning_ a hybrid cloud network, but rather how to run a cluster in the cloud and interact with an existing local network seamlessly.
+
+![Kubernetes in an hybrid cloud network](../images/kubernetes-hybrid-env.png)
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: `in a hybrid cloud network`

File: docs/kubernetes/hybrid-environment.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+- Deploying Kubernetes cluster in a [custom VNET using ACS-Engine](../../examples/vnet/README.md)
+- Azure [VPN Gateway](https://azure.microsoft.com/en-us/services/vpn-gateway/) and/or [Azure Express Route](https://azure.microsoft.com/en-us/services/expressroute/)
+- Azure [Virtual Network Peering](https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-peering-overview)
+
+## Network and Infrastructure
+
+### Topology
+
+The network topology must be well defined beforehand to enable peering between the different VNET. This means that the subnet ip range musty be defined before deploying kubernetes. It cannot be changed afterwards.
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: typo here `musty` --> `must`

File: docs/kubernetes/hybrid-environment.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+### Dns
+
+In a hybrid environment, you usually want to integrate with your on-premises DNS. There is two aspects to this. The first one is to register the VMs forming the cluster, and using your local search domain when resolving other services. The second is getting the services running on Kubernetes to use the external DNS.
+To benefit the scaling capabilities of the cluster and to ensure resiliency to machine failure, every node configuration needs to be scripted and part of the initial template that acs-engine will deploy. To register the nodes in your DNS at startup, you need to define [an acs-engine extension](../extensions.md) that will run your [dns registration script](https://github.com/tesharp/acs-engine/blob/register-dns-extension/extensions/register-dns/v1/register-dns.sh).
+
+In addition, you might want cluster services to address urls outside the cluster using your on-premise dns. To achieve this you need to configure KubeDNS to use your existing nameservice as upstream. [This setup is well documented on kubernetes blog](https://kubernetes.io/blog/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes)
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: `urls` --> `URLs`

Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: `dns` --> `DNS`

File: docs/kubernetes/hybrid-environment.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+By default, Kubernetes deployment with acs-engine expose the the admin api publicly (and securely). This can be avoided. Using peering with private/on-premise virtual network with ACS-Engine also allows you to create cloud-hosted [private cluster](features.md#private-cluster), with no endpoint exposed over the Internet.
+
+## Kubernetes Network
+
+For your kubernetes cluster to communicate with your on-premise network, you will have to deploy it to the existing vnet setup to communicate through your VPN/ExpressRoute (see Peering). Deploying to an existing VNET is documented under [Custom VNET](../custom-vnet.md).
+
+### Network
+
+Deploying Acs-engine on azure, you have 3 options of network policy. Azure CNI, Kubenet, or Calico.
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: `Acs-engine` --> `ACS-Engine`

File: docs/kubernetes/hybrid-environment.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+#### Azure CNI
+
+By default, acs-engine is using the [**azure cni** network policy](../../examples/networkpolicy/README.md#azure-container-networking-default) plugin. This has some advantages and some consequences that must be considered when defining the network where we deploy the cluster. CNI provides an integration with azure subnet ip addressing so that every pod created ny kubernetes is assigned an ip address from the corresponding subnet.
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: `every pod created ny  kubernetes`  --> `every pod created by kubernetes`

File: docs/kubernetes/hybrid-environment.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+- Plan for *large* address space.
+- Subnets where you deploy Kubernetes must be sized according to your scaling plan
+- You must account for [Kubernetes control plane](https://kubernetes.io/docs/concepts/overview/components/) services
+- Network Security must be applied at the subnet level, using Azure NSG
+- You can avoid masquerading on outgoing network calls (packets origin are the pod ip, not the node ip)
+
+#### Kubenet
+
+The built-in kubernetes network plugin is [Kubenet](https://kubernetes.io/docs/concepts/cluster-administration/network-plugins/#kubenet).
+Kubenet assigns virtual ips to the pods running in the cluster that are not part of the physical network infrastructure. The nodes are then configured to forward and masquerade the network calls using iptables rules. This means you can plan for a much smaller address space on your network as only the nodes will get an ip address.
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: Let's change all the occurrences of `ip` in the file to `IP` for consistency

File: docs/kubernetes/hybrid-environment.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+### Network
+
+Deploying Acs-engine on azure, you have 3 options of network policy. Azure CNI, Kubenet, or Calico.
+
+#### Azure CNI
+
+By default, acs-engine is using the [**azure cni** network policy](../../examples/networkpolicy/README.md#azure-container-networking-default) plugin. This has some advantages and some consequences that must be considered when defining the network where we deploy the cluster. CNI provides an integration with azure subnet ip addressing so that every pod created ny kubernetes is assigned an ip address from the corresponding subnet.
+All IP addresses are pre-allocated at provisionning time. By default, [acs-engine will pre-allocate 128 ips per node](https://github.com/Azure/azure-container-networking/blob/master/docs/acs.md#enabling-azure-vnet-plugins-for-an-acs-kubernetes-cluster) on the subnet.
+While this can be configured, new addresses will not be allocated dynamically. you need to plan for maximum scale.
```
Comment (CecileRobertMichon) [inline docs/kubernetes/hybrid-environment.md]: `you need to plan for maximum scale.` incomplete sentence?

**Comment by jcorioland** [comment]
@CecileRobertMichon thanks for the review. All typos are corrected now.

**Pull request merged**